modules/apps/client-extension/client-extension-web/src/main/resources/META-INF/resources/js/components/FileEntryPicker.tsx

```diff
@@ -0,0 +1,351 @@
+		try {
+			const response = await Liferay.Util.fetch(
+				'/o/com-liferay-frontend-token-definition-impl' + endpoint,
+				{
+					body,
+					method: 'POST',
+				}
+			);
+
+			const data = await response.json();
+
+			let displayType: DisplayType = 'success';
+
+			if (!response.ok) {
+				displayType = 'danger';
+
+				setFrontendTokenDefinitionRefValue(MALFORMED_JSON);
+			}
+
+			setFeedback({
+				displayType,
+				message: data.message,
+			});
+		}
+		catch (error) {
+			clearSelection();
+
+			setFeedback({
+				displayType: 'danger',
+				message: Liferay.Language.get(
+					'an-error-has-occurred-while-connecting-to-our-servers'
+				),
+			});
+		}
+	};
+
+	async function handleUpload(
+		file: File,
+		focusSelectFileButton: boolean = true
+	) {
+		const formData = new FormData();
+
+		formData.append('file', file);
+
+		await handlePost(formData, '/validate-file');
+
+		if (focusSelectFileButton) {
+			selectFileButtonRef.current?.focus();
+		}
+	}
+
+	const getClientExtensionFrontendTokenDefinitionInfo = () => {
+		handlePost(
+			Liferay.Util.objectToFormData({
+				companyId,
+				externalReferenceCode,
+			}),
+			'/tokens-info'
+		);
+	};
+
+	const updateFeedbackAfterSubmissionError = (
+		frontendTokenDefinition: string
+	) => {
+		handleUpload(
+			new File([frontendTokenDefinition], fileName, {
+				type: 'application/json',
+			}),
+			false
+		);
+	};
+
+	useEffect(() => {
+		if (!frontendTokenDefinitionRef.current) {
+			return;
+		}
+
+		frontendTokenDefinitionRef.current.value = frontendTokenDefinition;
+
+		if (frontendTokenDefinition && externalReferenceCode) {
+			getClientExtensionFrontendTokenDefinitionInfo();
+		}
+		else if (frontendTokenDefinition) {
+			updateFeedbackAfterSubmissionError(frontendTokenDefinition);
+		}
+		else {
+			clearSelection();
+		}
+
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+	}, []);
+
+	const buttonTitle = !fileName
+		? Liferay.Language.get('select-json')
+		: Liferay.Language.get('replace-json');
+
+	const inputId = `${namespace}file`;
+
+	return (
+		<>
+			<label
+				aria-describedby={
+					namespace + 'frontendTokenDefinitiondescription'
+				}
+				className="d-block"
+				htmlFor={inputId}
+				tabIndex={0}
+			>
+				{Liferay.Language.get('frontend-token-definition-json-upload')}
+			</label>
+
+			<Text
+				as="p"
+				color="secondary"
+				id={namespace + 'frontendTokenDefinitiondescription'}
+				size={3}
+			>
+				{Liferay.Language.get('frontend-token-definition-is-a-json')}{' '}
+				<a href="https://learn.liferay.com/w/dxp/site-building/site-appearance/style-books/developer-guide/style-book-token-definitions">
+					{Liferay.Language.get(
+						'learn-more-about-frontend-token-definition'
+					)}
+				</a>
+			</Text>
+
+			<ClayInput
+				accept=".json"
+				className="d-none"
+				id={inputId}
+				name={inputId}
+				onChange={onInputChange}
+				ref={inputFileRef}
+				type="file"
+			/>
+
+			<ClayInput
+				className="d-none"
```

Comment:
If this isn't going to be visible, can we do a hidden input instead of a textarea?

Reply:
Still needs to be applied.